app/geant-exporter/GeantPhysicsTableWriter.cc

```diff
@@ -0,0 +1,305 @@
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file GeantPhysicsTableWriter.cc
+//---------------------------------------------------------------------------//
+#include "GeantPhysicsTableWriter.hh"
+
+#include <fstream>
+
+#include <TFile.h>
+#include <TTree.h>
+#include <TBranch.h>
+
+#include <G4VProcess.hh>
+#include <G4VEmProcess.hh>
+#include <G4VEnergyLossProcess.hh>
+#include <G4VMultipleScattering.hh>
+#include <G4SystemOfUnits.hh>
+
+#include "GeantPhysicsTableWriterHelper.hh"
+#include "io/GeantPhysicsTable.hh"
+#include "io/GeantPhysicsVector.hh"
+#include "io/GeantTableType.hh"
+#include "io/GeantProcessType.hh"
+#include "io/GeantProcess.hh"
+#include "io/GeantPhysicsVectorType.hh"
+#include "io/GeantModel.hh"
+#include "base/Range.hh"
+#include "base/Types.hh"
+
+using celeritas::GeantModel;
+using celeritas::GeantPhysicsVector;
+using celeritas::GeantPhysicsVectorType;
+using celeritas::GeantProcess;
+using celeritas::GeantProcessType;
+using celeritas::GeantTableType;
+using celeritas::PDGNumber;
+using celeritas::real_type;
+using geant_exporter::to_geant_model;
+using geant_exporter::to_geant_physics_vector_type;
+using geant_exporter::to_geant_process;
+using geant_exporter::to_geant_process_type;
+using geant_exporter::to_geant_table_type;
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with existing TFile reference
+ */
+GeantPhysicsTableWriter::GeantPhysicsTableWriter(TFile* root_file)
+{
+    REQUIRE(root_file);
+    this->tree_tables_ = std::make_unique<TTree>("tables", "tables");
+    tree_tables_->Branch("GeantPhysicsTable", &(table_));
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Add physics tables to the ROOT file from given process and particle
+ */
+void GeantPhysicsTableWriter::add_physics_tables(
+    const G4VProcess& process, const G4ParticleDefinition& particle)
+{
+    // Process name
+    std::string process_name = process.GetProcessName();
+
+    // Write this->table_
+    table_.process_type = to_geant_process_type(process.GetProcessType());
+    table_.process      = to_geant_process(process_name);
+    table_.particle     = PDGNumber(particle.GetPDGEncoding());
+
+    if (auto em_process = dynamic_cast<const G4VEmProcess*>(&process))
+    {
+        // G4VEmProcess tables
+        this->fill_em_tables(*em_process);
+    }
+    else if (auto energy_loss
+             = dynamic_cast<const G4VEnergyLossProcess*>(&process))
+    {
+        // G4VEnergyLossProcess tables
+        this->fill_energy_loss_tables(*energy_loss);
+    }
+    else if (auto multiple_scattering
+             = dynamic_cast<const G4VMultipleScattering*>(&process))
```

Comment:
Just to double-check, but these G4V classes are all "siblings" in the hierarchy, so one of them doesn't override the other?

Reply:
What do you mean one of them overriding the other? Each process is going to have its own implementation of one of the 3 classes (em, eloss, or multiple scattering), so I don't see a problem (such as a discrete em process overriding a multiple scattering virtual method). 

Also, I probably should make them `const auto& em_process`, and so on. Same for all the `G4PhysicsTable` created in the `fill_blah_tables(...)`...

Reply:
I just meant from a class hierarchy: if `G4VMultipleScattering` is a subclass of `G4VEnergyLossProcess` then the `else` here means that it will never be called.

The `if` statements won't work if you change them from pointers to references -- `dynamic_cast` to an invalid reference raises an exception rather than returning a null pointer. The only change I'd make is to have `const auto*` rather than `auto` to be more explicit, although they end up the same.

Reply:
Both `G4VMultipleScattering` and `G4VEnergyLossProcess` are subclasses of `G4VContinuousDiscreteProcess` (great name, btw). So they are at the same hierarchical level.

Reply:
OK, so you should be good. My concern was that (for example) if you were to add a
```
else if (const auto* p = dynamic_cast<const G4VContinuousDiscreteProcess*>(&process))
```
above the MSC/ELoss `if`s, then those would be skipped (perhaps unintentionally).